### PR TITLE
[RFC] Signing TAs using subkeys

### PR DESCRIPTION
**Pull request by jenswi-linaro**
The pull request is a prototype to evaluate one way of having multiple signers for TAs.
See also https://linaro.atlassian.net/browse/TS-255

### CODE REVIEW

**Comment by jforissier** [comment]
For "core: add offset argument to shdr_alloc_and_copy()":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Review by jforissier — COMMENTED** [review]
One minor comment on "core: ree_fs: refactor check_update_version()", other than that LGTM.

**Review by jforissier — COMMENTED** [review]
A couple of comments on "core: support loading TAs signed with a subkey" below.

I need to better understand the UUID name scoping mechanism. Which purpose does it serve? What kind of bad things could occur without it? Why do subkeys need UUIDs in the first place?

Also, it would be cool to update `scripts/dump_ta_header.py`.

**Comment by jenswi-linaro** [comment]
In PR https://github.com/OP-TEE/optee_test/pull/615 there's a file, ta/subkey_notes.txt, which explains how the subkeys are used in the tests.

The usage in xtest isn't too far from what one might see in practice (except that all the subkeys obviously will not be kept together as that defeats the purpose with subkeys). For example the Top subkey is delegated to a organisation, the Mid key is delegated to a team in that organization. An identity subkey is used when a TA must use a specific UUID but we still need to delegate the signing.

File: core/kernel/ree_fs_ta.c [hunk 1]
```diff
@@ -67,13 +67,117 @@ struct ree_fs_ta_handle {
...
 	uint32_t db_version;
 	uint32_t nb_entries;
 };
 
-static const char ta_ver_db_obj_id[] = "ta_ver.db";
-static struct mutex ta_ver_db_mutex = MUTEX_INITIALIZER;
+static const char ta_ver_db[] = "ta_ver.db";
+static struct mutex ver_db_mutex = MUTEX_INITIALIZER;
+
+static TEE_Result check_update_version(const char *db_name,
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/kernel/ree_fs_ta.c:84]: OK, I'll update.

File: core/include/signed_hdr.h [hunk 2]
```diff
@@ -64,6 +65,62 @@ static inline size_t shdr_get_size(const struct shdr *shdr)
...
+ * @attr_count:		Number of attributes for the public key matching
+ *			@algo.
+ * @attrs:		Attributes for the public key matching @algo.
+ * @attrs[].id:		Attribute ID TEE_ATTR_* from GlobalPlatform
+ * @attrs[].offs:	Offset of the attribute value from start of
+ *			struct shdr_subkey
+ * @attrs[].size:	Attribute size
+ *
+ * The @uuid defines UUID URN Namespace (RFC4122), the next UUID after this
+ * header (another subkey or a TA) must be in the namespace of this UUID.
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/include/signed_hdr.h:86]: The idea with UUIDv5 is a security feature I borrowed from the GP Management Framework.
It's a pity that UUIDv5 doesn't support UUIDv5. This is mostly an internal matter to OP-TEE so we could break with the standard if we think it's worth it. If we do this we should probably have something in `struct shdr_subkey` to indicate this.

UUID derivation applies only to subkeys with a non-zero `name_size`. So existing TAs are not affected. However, if one would like sign an existing TA with a subkey there's the option with identity subkeys where `name_size` is zero and the UUID of the subkey is also used for the TA.

File: core/include/signed_hdr.h [hunk 3]
```diff
@@ -64,6 +65,62 @@ static inline size_t shdr_get_size(const struct shdr *shdr)
...
+ * hierarchy and cannot be moved up or below another subkey.
+ *
+ * If @name_size is non-zero it indicates that a name field of this size
+ * exists and is used to generate the UUID of the following TA or subkey.
+ * If it's zero the following TA or subkey must have a matching UUID.
+ *
+ * The @subkey_version field is used as a rollback measure. The version is
+ * checked against earlier saved values of this subkey. If the latest known
+ * version is less than this the stored value is update. If the latest
+ * known version is larger than this then the subkey is refused.
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/include/signed_hdr.h:97]: Yes, key revocation.

**Comment by jenswi-linaro** [comment]
> I need to better understand the UUID name scoping mechanism. Which purpose does it serve? What kind of bad things could occur without it? Why do subkeys need UUIDs in the first place?

The purpose with UUID name scoping is to avoid clashes of TA UUIDs signed with different subkeys. If that should happen then the owner of subkey1 could sign TAs impersonating TAs signed by subkey2.

Subkeys has UUIDs because it's convenient way of starting a new name space. I could have chosen something different to generate the TA UUID with, but I wanted something that can be kept constant even if the public key in the subkey is replaced. The constant is need to avoid changing UUID of all TAs signed with a certain private key. Let's say that subkey1 changes owner, it would be nice to be able to change keypair without changing UUID of all TAs that follows with this subkey.

**Comment by jenswi-linaro** [comment]
> Also, it would be cool to update `scripts/dump_ta_header.py`.

I had missed that script. `scripts/sign_encrypt.py parse` does more or less the same thing in this PR.
Should I remove the parse command from `scripts/sign_encrypt.py` and update `scripts/dump_ta_header.py` instead?

**Comment by jforissier** [comment]
> In PR [OP-TEE/optee_test#615](https://github.com/OP-TEE/optee_test/pull/615) there's a file, ta/subkey_notes.txt, which explains how the subkeys are used in the tests.

Cool.
 
> The usage in xtest isn't too far from what one might see in practice (except that all the subkeys obviously will not be kept together as that defeats the purpose with subkeys). For example the Top subkey is delegated to a organisation, the Mid key is delegated to a team in that organization. An identity subkey is used when a TA must use a specific UUID but we still need to delegate the signing.

Why "identity"? Because the UUID of the key is the same as the TA UUID?

**Comment by jforissier** [comment]
> > Also, it would be cool to update `scripts/dump_ta_header.py`.
> 
> I had missed that script. `scripts/sign_encrypt.py parse` does more or less the same thing in this PR. Should I remove the parse command from `scripts/sign_encrypt.py` and update `scripts/dump_ta_header.py` instead?

No, I totally missed the `parse`command. In fact it's probably better than having a separate script especially if we want to be able to perform any kind of verification.

**Comment by jenswi-linaro** [comment]
> > The usage in xtest isn't too far from what one might see in practice (except that all the subkeys obviously will not be kept together as that defeats the purpose with subkeys). For example the Top subkey is delegated to a organisation, the Mid key is delegated to a team in that organization. An identity subkey is used when a TA must use a specific UUID but we still need to delegate the signing.
> 
> Why "identity"? Because the UUID of the key is the same as the TA UUID?

Yes, I thought that would be intuitive. I'm open to a better term.

**Comment by jenswi-linaro** [comment]
Replaced by https://github.com/OP-TEE/optee_os/pull/5584

**Pull request closed without merging**